### PR TITLE
feat(bootstrap): support customizing corefile template from kuma-cp

### PR DESCRIPTION
**Pull request by jijiechen**
### Checklist prior to review

- [x] [Link to relevant issue][1] as well as docs and UI issues
  - https://github.com/kumahq/kuma/issues/8119
- [x] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS
  - Should not block as it's adding a new feature.
- [x] Tests (Unit test, E2E tests, manual test on universal and k8s)
  - Added
  - Don't forget `ci/` labels to run additional/fewer tests
    - Should run normally
- [x] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? 
  - No, it's compatible with old versions
- [x] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) 
  - No backport needed

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

**Comment by lahabana** [comment]
No need for the label :)

File: app/kuma-dp/cmd/run_test.go [hunk 1]
```diff
@@ -116,35 +121,22 @@ var _ = Describe("run", func() {
...
-			Expect(actualArgs[1]).To(Equal("--config-path"))
-			actualConfigFile := actualArgs[2]
-			Expect(actualConfigFile).To(BeARegularFile())
-
-			// then
-			if given.expectedFile != "" {
-				Expect(actualArgs[2]).To(Equal(given.expectedFile))
-			}
+			corednsPid := verifyComponentProcess("coredns", corednsPidFile, corednsCmdlineFile, func(actualArgs []string) {
+				Expect(len(actualArgs)).To(HaveLen(3))
```
[the remark this answers is not in the document]
Comment (slonka) [reply app/kuma-dp/cmd/run_test.go:136]: you need to pass array to `HaveLen`

File: app/kuma-dp/cmd/run.go [hunk 2]
```diff
@@ -226,6 +203,37 @@ func newRunCmd(opts kuma_cmd.RunCmdOpts, rootCtx *RootContext) *cobra.Command {
 			}
 			opts.AdminPort = bootstrap.GetAdmin().GetAddress().GetSocketAddress().GetPortValue()
 
+			if cfg.DNS.Enabled && !cfg.Dataplane.IsZoneProxy() {
+				if cfg.DNS.CoreDNSConfigTemplatePath == "" && len(kumaSidecarConfiguration.Networking.CorefileTemplate) > 0 {
+					templateFile, err := writeTempFile("kuma-dp-corefile-template-", "Corefile", kumaSidecarConfiguration.Networking.CorefileTemplate, 0o644)
+					if err != nil {
+						return errors.Errorf("Failed to write corefile template to disk. %v", err)
```
Comment (jakubdyszkiewicz) [inline app/kuma-dp/cmd/run.go:210]: this is "nit", but errors shoudl be lowercase `failed to write...`. Check all errors in this PR for it.

**Review by jakubdyszkiewicz — APPROVED** [review]
Nice contrib!

Can you also add this new setting to `pkg/config/app/kuma-cp/kuma-cp.defaults.yaml` ?

File: app/kuma-dp/cmd/run.go (hunk 2 above)
Comment (jijiechen) [reply app/kuma-dp/cmd/run.go:210]: Thanks, will update code in an upcoming commit.

**Comment by jijiechen** [comment]
I'm still working on passing the E2E tests

**Comment by jijiechen** [comment]
The E2E tests are passing.

File: test/e2e_env/kubernetes/bootstrap/corefile_template.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
+package bootstrap
```
Comment (jakubdyszkiewicz) [inline test/e2e_env/kubernetes/bootstrap/corefile_template.go:1]: The test is nice, but I don't think this should go into `e2e_env/kubernetes`.

The idea of `test/e2e_env/kubernetes` is to be able to parallelize tests. This test breaks this promise. Yes, we use Kuma2 here, but imagine we also have another test that needs to deploy Kuma with a custom config using Kuma2. It cannot be then used in parallel with this test.

Another disadvantage is that we need to run a second Kube cluster just for this test. When developing a kube e2e test, usually you focus the suite you are writing and run `make test/e2e-kubernetes`. Forcing running a second kube cluster every single time prolongs the development loop.

I'd move this to `test/e2e`.

Comment (jijiechen) [reply test/e2e_env/kubernetes/bootstrap/corefile_template.go:1]: Yeah, that'll make more sense, you may go ahead moving it and revert the changes in `e2e.mk`. Thanks a lot.
I wasn't sure if it's good to place it under `e2e_env/kubernetes`, but I felt like it's also not very suitable to but in `test/e2e ` when compared to existing tests.

Comment (jijiechen) [reply test/e2e_env/kubernetes/bootstrap/corefile_template.go:1]: A new PR has been created to conduct this change, please review:
https://github.com/kumahq/kuma/pull/8736

**Pull request merged**